src/lines_to_params.rs

```diff
@@ -0,0 +1,165 @@
+
+type StatuslistParams<'a> = (
+    (u32, &'a str),
+    &'a str,
+    Vec<u32>,
+    Vec<f32>,
+    Vec<(u16, u16, f32, u32)>,
+);
+
+/// Get id name type from tuple
+#[pyfunction]
+#[pyo3(text_signature = "(name_id_pair: list[str]) -> (int, str)")]
+pub(crate) fn parse_id_name_pair(inp: Vec<&str>) -> (u32, &str) {
+    (
+        parser::u32_from_param(inp.first().unwrap()),
+        inp.last().unwrap(),
+    )
+}
+
+/// Params to ability
+#[pyfunction]
+#[pyo3(text_signature = "(timestamp: int, params: list[str]) -> Ability")]
+pub(crate) fn ability_from_params(timestamp: i64, inp: Vec<&str>) -> PyObject {
+    let mut col = inp;
+    let source_actor = col.drain(..2).collect::<Vec<&str>>();
+    let ability = col.drain(..2).collect::<Vec<&str>>();
+    let target_actor = col.drain(..2).collect::<Vec<&str>>();
+    let action_effects = col.drain(..16).collect::<Vec<&str>>();
+    let source_resources = col.drain(..6)
+        .map(|x| parser::u32_from_str(x))
+        .collect::<Vec<u32>>();
+    let source_position = col.drain(..4)
+        .map(|x| parser::f32_from_str(x))
+        .collect::<Vec<f32>>();
+    let target_resources = col.drain(..6)
+        .map(|x| parser::u32_from_str(x))
+        .collect::<Vec<u32>>();
+    let target_position = col.drain(..4)
+        .map(|x| parser::f32_from_str(x))
+        .collect::<Vec<f32>>();
+    let sequence = parser::u32_from_param(col.first().unwrap());
+    Python::with_gil(|py| {
+        let source_actor_obj = create_id_name_pair(py, "nari.types.actor", "Actor", source_actor);
+        let target_actor_obj = create_id_name_pair(py, "nari.types.actor", "Actor", target_actor);
+        update_position(source_actor_obj, source_position);
+        update_position(target_actor_obj, target_position);
+        update_resources(source_actor_obj, source_resources);
+        update_resources(target_actor_obj, target_resources);
+
+        let ability_event = create_id_name_pair(py, "nari.types.ability", "Ability", ability);
+
+        let action_effect = PyModule::import(py, "nari.types.actioneffect").unwrap()
+            .getattr("ActionEffect").unwrap();
+
+        let action_effects = action_effects
+            .chunks(2)
+            .map(|x| {
+                let (param0, param1, severity, effect_category, value, flags, multiplier) = action_effect_from_params(x.to_vec());
+                let kwargs = PyDict::new(py);
+                kwargs.set_item("effect_category", effect_category).unwrap();
+                kwargs.set_item("severity", severity).unwrap();
+                kwargs.set_item("flags", flags).unwrap();
+                kwargs.set_item("value", value).unwrap();
+                kwargs.set_item("multiplier", multiplier).unwrap();
+                kwargs.set_item("additional_params", (param0, param1)).unwrap();
+                action_effect.call((), Some(kwargs)).unwrap()
+            }).collect::<Vec<&PyAny>>();
+
+        let kwargs = PyDict::new(py);
```

Comment:
Why not just instantiate the class... I'm unclear why this function is messing with so many Py things in general tho, it didn't last time I looked.

Reply:
It's so we could have a nicer exposed function return instead of `tuple[tuple[int, str], list[int], list[float], tuple[int, str], list[int], list[float], tuple[int, str], list[list[str]], int]` we have `Ability`

Reply:
We should do it properly in that case.

Reply:
What would you say is the proper way. Having a large tuple export or a clean single class export? @nonowazu maybe you have an opinion on this as well.